programs/bpf_loader/src/serialization.rs

```diff
@@ -175,67 +176,77 @@ pub fn serialize_parameters(
     instruction_context: &InstructionContext,
     should_cap_ix_accounts: bool,
     copy_account_data: bool,
-) -> Result<(AlignedMemory<HOST_ALIGN>, Vec<MemoryRegion>, Vec<usize>), InstructionError> {
+) -> Result<
+    (
+        AlignedMemory<HOST_ALIGN>,
+        Vec<MemoryRegion>,
+        Vec<SerializedAccountMetadata>,
+    ),
+    InstructionError,
+> {
     let num_ix_accounts = instruction_context.get_number_of_instruction_accounts();
     if should_cap_ix_accounts && num_ix_accounts > MAX_INSTRUCTION_ACCOUNTS as IndexOfAccount {
         return Err(InstructionError::MaxAccountsExceeded);
     }
-    let is_loader_deprecated = *instruction_context
-        .try_borrow_last_program_account(transaction_context)?
-        .get_owner()
-        == bpf_loader_deprecated::id();
 
-    let num_accounts = instruction_context.get_number_of_instruction_accounts() as usize;
-    let mut accounts = Vec::with_capacity(num_accounts);
-    let mut account_lengths: Vec<usize> = Vec::with_capacity(num_accounts);
-    for instruction_account_index in 0..instruction_context.get_number_of_instruction_accounts() {
-        if let Some(index) =
-            instruction_context.is_instruction_account_duplicate(instruction_account_index)?
-        {
-            accounts.push(SerializeAccount::Duplicate(index));
-            // unwrap here is safe: if an account is a duplicate, we must have
-            // seen the original already
-            account_lengths.push(*account_lengths.get(index as usize).unwrap());
-        } else {
-            let account = instruction_context
-                .try_borrow_instruction_account(transaction_context, instruction_account_index)?;
-            account_lengths.push(account.get_data().len());
-            accounts.push(SerializeAccount::Account(
-                instruction_account_index,
-                account,
-            ));
-        };
-    }
+    let (program_id, is_loader_deprecated) = {
+        let program_account =
+            instruction_context.try_borrow_last_program_account(transaction_context)?;
+        (
+            *program_account.get_key(),
+            *program_account.get_owner() == bpf_loader_deprecated::id(),
+        )
+    };
+
+    let accounts = (0..instruction_context.get_number_of_instruction_accounts())
+        .map(|instruction_account_index| {
+            if let Some(index) = instruction_context
+                .is_instruction_account_duplicate(instruction_account_index)
+                .unwrap()
```

Comment:
(note to self) seems this and below new `.unwrap()`s are safe. i.e. these are always infallible before/after this pr.